src/uct/ib/rc/accel/gga_mlx5.c

```diff
@@ -187,6 +227,205 @@ uct_gga_mlx5_rkey_resolve(uct_ib_mlx5_md_t *md, uct_rkey_t rkey)
 
 static UCS_CLASS_DECLARE_DELETE_FUNC(uct_gga_mlx5_iface_t, uct_iface_t);
 
+static ucs_status_t
+uct_gga_mlx5_iface_event_fd_get(uct_iface_h tl_iface, int *fd_p)
+{
+    uct_rc_mlx5_iface_common_t *iface =
+            ucs_derived_of(tl_iface, uct_rc_mlx5_iface_common_t);
+    uct_ib_mlx5_md_t *md UCS_V_UNUSED =
+            uct_ib_mlx5_iface_md(&iface->super.super);
+
+    ucs_assert(md->flags & UCT_IB_MLX5_MD_FLAG_DEVX_CQ);
+    *fd_p = iface->cq_event_channel->fd;
+    return UCS_OK;
+}
+
+static ucs_status_t
+uct_gga_mlx5_iface_arm(uct_iface_h tl_iface, unsigned events)
+{
+    uct_rc_mlx5_iface_common_t *iface =
+            ucs_derived_of(tl_iface, uct_rc_mlx5_iface_common_t);
+    uct_ib_mlx5_md_t *md UCS_V_UNUSED =
+            uct_ib_mlx5_iface_md(&iface->super.super);
+
+    ucs_assert(md->flags & UCT_IB_MLX5_MD_FLAG_DEVX_CQ);
+    return uct_rc_mlx5_iface_devx_arm(iface, events);
+}
+
+static ucs_status_t
+uct_gga_mlx5_iface_query(uct_iface_h tl_iface, uct_iface_attr_t *iface_attr)
+{
+    uct_ib_iface_t *iface = ucs_derived_of(tl_iface, uct_ib_iface_t);
+    ucs_status_t status;
+
+    status = uct_ib_iface_query(iface, UCT_IB_RETH_LEN, iface_attr);
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    iface_attr->device_addr_len += ucs_offsetof(uct_gga_mlx5_dev_addr_t,
+                                                ib_addr);
+    iface_attr->ep_addr_len      = sizeof(uct_gga_mlx5_ep_address_t);
+    iface_attr->cap.flags = /*
+                            UCT_IFACE_FLAG_PUT_ZCOPY |
+                            UCT_IFACE_FLAG_GET_ZCOPY |
+                            UCT_IFACE_FLAG_PENDING   | */
+                            UCT_IFACE_FLAG_CONNECT_TO_EP |
+                            UCT_IFACE_FLAG_CB_SYNC;
+
+    iface_attr->cap.event_flags = UCT_IFACE_FLAG_EVENT_SEND_COMP |
+                                  UCT_IFACE_FLAG_EVENT_FD;
+    return UCS_OK;
+}
+
+static ucs_status_t
+uct_gga_mlx5_ep_enable_mmo(uct_gga_mlx5_ep_t *ep)
+{
+    uct_rc_mlx5_iface_common_t *iface = ucs_derived_of(ep->super.super.super.super.iface,
+                                                       uct_rc_mlx5_iface_common_t);
+    uct_ib_mlx5_md_t *md              = ucs_derived_of(iface->super.super.super.md,
+                                                       uct_ib_mlx5_md_t);
+
+    char in[UCT_IB_MLX5DV_ST_SZ_BYTES(init2init_qp_in)] = {};
+    char out[UCT_IB_MLX5DV_ST_SZ_BYTES(init2init_qp_out)] = {};
+    void *qpce = UCT_IB_MLX5DV_ADDR_OF(init2init_qp_in, in, qpc_data_extension);
+    int rc;
+
+    /* TODO: built-in to EP */
+    rc = ucs_posix_memalign((void**)&ep->dma_opaque.buf,
+                            sizeof(uct_gga_mlx5_dma_opaque_t),
+                            sizeof(uct_gga_mlx5_dma_opaque_t), "gga_opaque_buf");
+    if (rc != 0) {
+        ucs_error("cannot allocate MMO opaque buffer: %m");
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    ep->dma_opaque.mr = ibv_reg_mr(md->super.pd, ep->dma_opaque.buf,
```

Comment:
should we reuse here uct_ib_reg_mr?
uct_ib_reg_mr handles EAGAIN error by repeating the call, should it be the similar in this case?

Reply:
> should we reuse here uct_ib_reg_mr?

IMO it's overcomplicated in this case
> uct_ib_reg_mr handles EAGAIN error by repeating the call, should it be the similar in this case?

no, that's only for ODP case